brush-core/src/builtins/mapfile.rs

```diff
@@ -93,46 +80,71 @@ impl builtins::Command for MapFileCommand {
 }
 
 impl MapFileCommand {
-    fn read_entries(
+    async fn read_entries(
         &self,
-        mut input_file: openfiles::OpenFile,
+        input_file: openfiles::OpenFile,
+        context: &mut commands::ExecutionContext<'_>,
     ) -> Result<variables::ArrayLiteral, error::Error> {
-        let mut entries = vec![];
-
         let orig_term_attr = setup_terminal_settings(&input_file)?;
 
-        let mut current_entry = String::new();
-        let mut buffer: [u8; 1] = [0; 1]; // 1-byte buffer
+        let mut entries = vec![];
+        let mut line = vec![];
+        let mut idx = self.skip_count;
+        let mut read_count = 0;
+        let delimiter = self.delimiter.chars().next().unwrap_or('\n') as u8;
+
+        let mut reader = input_file.clone();
+        let mut buf = [0u8; 1];
+
+        while self.max_count == 0 || entries.len() < usize::try_from(self.max_count)? {
+            line.clear();
+            let mut saw_delimiter = false;
+
+            loop {
+                match reader.read(&mut buf) {
+                    Ok(0) => break,                                         // End of input
+                    Ok(1) if buf[0] == b'\x03' => break,                    // Ctrl+C
+                    Ok(1) if buf[0] == b'\x04' && line.is_empty() => break, // Ctrl+D
+                    Ok(1) => {
+                        let byte = buf[0];
+                        line.push(byte);
+                        if byte == delimiter {
+                            saw_delimiter = true;
+                            break;
+                        }
+                    }
+                    Ok(_) => unreachable!("input can only be 0, 1, or error"),
+                    Err(e) => return Err(e.into()),
+                }
+            }
 
-        loop {
-            // TODO: Figure out how to restore terminal settings on error?
-            let n = input_file.read(&mut buffer)?;
-            if n == 0 {
-                // EOF reached.
+            if line.is_empty() && !saw_delimiter {
                 break;
             }
 
-            let ch = buffer[0] as char;
+            if read_count < self.skip_count {
```

Comment:
Minor suggestion: right now we're tracking `read_count` and `idx`. Can we ditch `read_count` and just use `idx` as long as we (1) initialize `idx` to `0` (instead of `skip_count`) and (2) make sure to increment it even for skipped reads (e.g., right before line 125)? I think that will be somewhat more readable.